File: lilguys/snapshot.py
import h5py
import numpy as np
from os import path

from .hdfutils import get_h5_vector, get_h5_header, set_h5_vector, set_h5_header, make_default_header
from . import units


class Snapshot:
    _filename = None

    def __init__(self, positions, velocities, 
            IDs=None, potential=None, accelerations=None,
            header = {}, m=None,
            ext_potential=None):

        self.pos = positions
        self.vel = velocities

        if IDs is None:
            IDs = np.arange(len(self))
        if potential is None:
            potential = np.zeros(len(self))
        if ext_potential is None:
            ext_potential = np.zeros(len(self))

        self.acc = accelerations
        self.IDs = IDs
        self.potential = potential
        self.ext_potential = ext_potential
        if header == {}:
            N = len(self)
            header = make_default_header(N, m)
        self.header = header

    @classmethod
    def file(cls, filename):
        with h5py.File(filename, "r") as f:
            keys = f["PartType1"].keys()
            header = get_h5_header(f)

            pos = get_h5_vector(f, "Coordinates")
            vel = get_h5_vector(f, "Velocities")
            IDs = get_h5_vector(f, "ParticleIDs")

            if "Acceleration" in keys:
                acc = get_h5_vector(f, "Acceleration")
            else:
                acc = None
            if "Potential" in keys:
                pot = get_h5_vector(f, "Potential")
            else:
                pot = None

            if "ExtPotential" in keys:
                ext_pot = get_h5_vector(f, "ExtPotential")
            else:
                ext_pot = None

        c = cls(pos, vel, IDs=IDs, header=header, potential=pot, 
                accelerations=acc, ext_potential=ext_pot)
        c._filename = filename
        return c


    def save(self, filename=None):
        if filename is None and path.exists(self._filename):
            self._save()
            print("snapshot saved")
        else:
            if filename is not None:
                self._filename = filename
            self._save_new()
            print("snapshot saved at ", self._filename)
        return True


    def filter(self, filt, inplace=False):
        assert len(filt) == len(self)
        if inplace:
            copy = self
        else:
            copy = self.copy()
        copy.pos = copy.pos[filt]
        copy.vel = copy.vel[filt]
        copy.IDs = copy.IDs[filt]
        copy.header["NumPart_ThisFile"] = (0, len(copy))
        copy.header["NumPart_Total"] = (0, len(copy))
        if copy.potential is not None:
            copy.potential = copy.potential[filt]
        return copy

    def _save(self):

        with h5py.File(self._filename, "a") as f:
            if self.gadget4:
                set_h5_vector(f, "Position", self.pos)
                set_h5_vector(f, "Velocity", self.vel)
            else:
                set_h5_vector(f, "Coordinates", self.pos)
                set_h5_vector(f, "Velocities", self.vel)

            set_h5_vector(f, "ParticleIDs", self.IDs)
            set_h5_header(f, self.header)

    def _save_new(self):
        with h5py.File(self._filename, "w") as f:
            part = f.create_group("PartType1")
            f.create_group("Header")

            set_h5_vector(f, "ParticleIDs", self.IDs)
            set_h5_vector(f, "Coordinates", self.pos)
            set_h5_vector(f, "Velocities", self.vel)
            set_h5_vector(f, "Potential", self.potential)
            set_h5_vector(f, "Acceleration", np.zeros(len(self)))
            set_h5_header(f, self.header)

    def calc_potential(self, h=0.08):
        for i in range(len(self)):
            p1 = self.pos[i]
            p2 = self.pos
            r = np.sqrt(np.sum((p1 - p2)**2, -1))
            self.potential[i] = -units.G * self.m * np.sum(1/np.sqrt(r**2 + h**2))



    def shift(self, p0, v0=[0,0,0], inplace=False):
        if inplace:
            self.pos += p0
            self.vel += v0
            return self
        else:
            shifted = self.copy()
            shifted.pos += p0
            shifted.vel += v0
            return shifted

    def scale(self, r_scale=1, v_scale=1, m_scale=1, inplace=False):
        if inplace:
            self.pos *= r_scale
            self.vel *= v_scale
            self.m *= m_scale
        else:
            scaled = self.copy()
            scaled.pos *= r_scale
            scaled.vel *= v_scale
            scaled.m *= m_scale
            return scaled


    def copy(self):
        pos = np.copy(self.pos)
        vel = np.copy(self.vel)
        IDs = np.copy(self.IDs)
        potential = np.copy(self.potential)
        header = self.header.copy()
        return Snapshot(pos, vel, IDs=IDs, potential=potential, header=header)


    @property
    def x(self):
        return self.pos[:,0]

    @property
    def y(self):
        return self.pos[:,1]


    @property
    def z(self):
        return self.pos[:,2]

    @property
    def v_x(self):
        return self.vel[:,0]

    @property
    def v_y(self):
        return self.vel[:,1]

    @property
    def v_z(self):
        return self.vel[:,2]

    @property
    def r(self):
        return np.sqrt(self.x**2 + self.y**2 + self.z**2)

    @property
    def v(self):
        return np.sqrt(self.v_x**2 + self.v_y**2 + self.v_z**2)


    @property
    def m(self):
        return self.header["MassTable"][1]

    @m.setter
    def m(self, a):
        self.header["MassTable"][1] = a

    def __str__(self):
        return f"snapshot with {len(self)} particles"

    def __repr__(self):
        return str(self)

    def __len__(self):
        return len(self.x)

File: LilGuys/src/snapshot.jl
using LinearAlgebra: norm
using Printf
import Base: @kwdef


"""A map of the HDF5 columns to the snapshot fields"""
const h5vectors = Dict(
    :Φs=>"Potential",
    :Φs_ext=>"ExtPotential",
    :index=>"ParticleIDs",
    :positions=>"Coordinates",
    :velocities=>"Velocities",
    :accelerations=>"Acceleration",
    :masses=>"Masses",
   )

const snap_matricies = [:positions, :velocities, :accelerations]
const snap_vectors = [:masses, :Φs, :Φs_ext, :index]



"""
A snapshot of a gadget simulation. Units are all code units.
"""
@kwdef mutable struct Snapshot 
    """The positions of the particles"""
    positions::Matrix{F}

    """The velocities of the particles"""
    velocities::Matrix{F}

    """The masses of the particles"""
    masses::Union{Vector, ConstVector}

    """The particle IDs"""
    index::Vector{Int}  

    """The accelerations of the particles"""
    accelerations::OptMatrix = nothing

    """The potential of the particles"""
    Φs::OptVector = nothing  

    """The external (milky way) potential of the particles"""
    Φs_ext::OptVector = nothing

    """The filename of the snapshot"""
    filename::String = ""

    """softening length"""
    h::Real = NaN

    """The Gadget HDF5 header"""
    header::Dict{String, Any}

    """The (adopted) centre"""
    x_cen::Vector{F} = zeros(F, 3)

    """The (adopted) velocity centre"""
    v_cen::Vector{F} = zeros(F, 3)

    """The (stellar) weights of the particles"""
    weights::Union{Vector, ConstVector} = ConstVector(1.0, 0)

    """The radii of the particles, stored on first calculation"""
    _radii::OptVector = nothing
end




"""
    Snapshot(positions, velocities, mass::Real)

Create a snapshot with constant mass.
"""
function Snapshot(positions, velocities, mass::Real)
    N = size(positions, 2)
    masses = ConstVector(mass, N)
    return Snapshot(positions, velocities, masses)
end


"""
    Snapshot(positions, velocities, masses)

Create a snapshot.
"""
function Snapshot(positions, velocities, masses)
    N = size(positions, 2)
    if mass_is_fixed(masses)
        m_header = masses[1]
    else
        m_header = 0.0
    end

    header = make_default_header(N, m_header)
    index = collect(1:N)
    return Snapshot(positions=positions, velocities=velocities, masses=masses, header=header, index=index)
end



"""
    Snapshot(filename)

Load a snapshot from an HDF5 file.
"""
function Snapshot(filename::String)
    h5open(filename, "r") do h5f
        return Snapshot(h5f, filename=filename)
    end
end


"""
    Snapshot(h5f; mmap=false, filename)

Load a snapshot from an HDF5 file.
"""
function Snapshot(h5f::HDF5.H5DataStore; mmap=false, filename="", group="PartType1")
    kwargs = Dict{Symbol, Any}()

    for (var, col) in h5vectors
        if col ∈ keys(h5f[group])
            kwargs[var] = get_vector(h5f, "$group/$col", mmap=mmap)
        end
    end

    header = get_header(h5f)
    kwargs[:header] = header
    m = header["MassTable"][2]
    N = header["NumPart_ThisFile"][2]

    if m != 0  || (m == 0 && :masses ∉ keys(kwargs))
        kwargs[:masses] = ConstVector(m, N)
    end

    kwargs[:filename] = filename

    return Snapshot(; kwargs...)
end


function Base.size(snap::Snapshot) 
    return (length(snap.index),)
end


function Base.length(snap::Snapshot) 
    return length(snap.index)
end

# Base.IndexStyle(::Type{<:Snapshot}) = IndexLinear()


function Base.getindex(snap::Snapshot, idx)
    kwargs = Dict{Symbol, Any}()
    kwargs[:h] = snap.h
    kwargs[:x_cen] = snap.x_cen
    kwargs[:v_cen] = snap.v_cen
    for sym in fieldnames(Snapshot)
        if getproperty(snap, sym) === nothing
            continue
        elseif !isa(idx, Int) && sym ∈ [:header, :filename]
            kwargs[sym] = getproperty(snap, sym)
        elseif sym ∈ snap_matricies
            kwargs[sym] = getproperty(snap, sym)[:, idx]
        elseif sym ∈ snap_vectors
            kwargs[sym] = getproperty(snap, sym)[idx]
        elseif sym ∈ [:weights]
            if isa(getproperty(snap, sym), ConstVector)
                kwargs[sym] = getproperty(snap, sym)
            else
                kwargs[sym] = getproperty(snap, sym)[idx]
            end
        else
            continue
        end
    end

    return Snapshot(; kwargs...)
end



"""is the particle mass constant in the snapshot?"""
function mass_is_fixed(snap::Snapshot)
    return mass_is_fixed(snap.masses)
end


function mass_is_fixed(masses::Union{Vector, ConstVector})
    return masses isa ConstVector && masses[1] != 0 || all(masses .== masses[1])
end


"""
    save(filename, snap)

Save a snapshot to an HDF5 file.
"""
function save(filename::String, snap::Snapshot)
    h5open(filename, "w") do h5f
        save(h5f, snap)
    end
end


function save(snap::Snapshot)
    save(snap.filename, snap)
end



function save(h5f::HDF5.H5DataStore, snap::Snapshot)
    if mass_is_fixed(snap)
        snap.header["MassTable"][2] = snap.masses[1]
    end
    set_header!(h5f, snap.header)

    for (var, _) in h5vectors
        save_vector(h5f, snap, var)
    end
end


"""
saves a vector from a snapshot into an HDF5 file
"""
function save_vector(h5f::HDF5.H5DataStore, snap::Snapshot, var::Symbol; group="PartType1")
    col = h5vectors[var]
    val = getproperty(snap, var) 

    if group ∉ keys(h5f)
        HDF5.create_group(h5f, group)
    end

    if var == :masses && mass_is_fixed(snap)
        # pass
    elseif val !== nothing
        set_vector!(h5f, "$group/$col", val)
    end
end



function Base.show(io::IO, snap::Snapshot)
    print(io, "<snapshot with $(length(snap)) particles>")
    return io
end



"""
    make_default_header(N, mass)

Create a default Gadget header for a snapshot with N particles and DM mass `mass`.
"""
function make_default_header(N, mass)

    return Dict(
        "NumPart_ThisFile"=>F[0, N],
        "MassTable"=>F[0.0, mass],
        "Time"=>0.0,
        "Redshift"=>0.0,
        "NumPart_Total"=>F[0, N],
        "NumFilesPerSnapshot"=>1,
        "BoxSize"=>1000,
       )

end


"""
Regenerates the header of the snapshot.
"""
function regenerate_header!(snap::Snapshot)
    if mass_is_fixed(snap)
        m = snap.masses[1]
    else
        m = 0.0
    end

    N = length(snap)
    snap.header = make_default_header(N, m)
end


"""
    make_gadget2_header(N, mass)

Create a Gadget-2 header for a snapshot with N particles and DM mass `mass`.
"""
function make_gadget2_header(N, mass)
    return Dict(
        "NumPart_ThisFile"=>F[0, N, 0, 0, 0, 0],
        "MassTable"=>F[0.0, mass, 0.0, 0.0, 0.0, 0.0],
        "Time"=>0.0,
        "Redshift"=>0.0,
        "NumPart_Total"=>F[0, N, 0, 0, 0, 0],
        "NumFilesPerSnapshot"=>1,
        "Flag_Entropy_ICs"=>0,
        "NumPart_Total_HighWord"=>F[0, 0, 0, 0, 0, 0],
    )
end


